module/zfs/zfs_ioctl.c

```diff
@@ -2819,8 +2777,49 @@ zfs_ioc_inherit_prop(zfs_cmd_t *zc)
 			return (SET_ERROR(EINVAL));
 	}
 
-	/* property name has been validated by zfs_secpolicy_inherit_prop() */
-	return (dsl_prop_inherit(zc->zc_name, zc->zc_value, source));
+	if (prop == ZPROP_INVAL) {
+		if (!zfs_prop_user(propname))
+			return (SET_ERROR(EINVAL));
+
+		type = PROP_TYPE_STRING;
+	} else if (prop == ZFS_PROP_VOLSIZE || prop == ZFS_PROP_VERSION) {
+		return (SET_ERROR(EINVAL));
+	} else {
+		type = zfs_prop_get_type(prop);
+	}
+
+	/*
+	 * zfs_prop_set_special() expects properties in the form of an
+	 * nvpair with type info.
+	 */
+	dummy = fnvlist_alloc();
+
+	switch (type) {
+	case PROP_TYPE_STRING:
+		VERIFY(0 == nvlist_add_string(dummy, propname, ""));
+		break;
+	case PROP_TYPE_NUMBER:
+	case PROP_TYPE_INDEX:
+		VERIFY(0 == nvlist_add_uint64(dummy, propname, 0));
+		break;
+	default:
+		err = SET_ERROR(EINVAL);
+		goto errout;
+	}
+
+	pair = nvlist_next_nvpair(dummy, NULL);
+	if (pair == NULL) {
+		err = SET_ERROR(EINVAL);
+	} else {
+		err = zfs_prop_set_special(zc->zc_name, source, pair);
+		if (err == -1) /* property is not "special", needs handling */
+			err = dsl_prop_inherit(zc->zc_name, zc->zc_value,
```

Comment:
@loli10K 

yes, a simple way to control the overheads is to introduce a `PROP_SPECIAL` flag to the `zfs_prop_table`,  that could be checked for (`boolean_t zfs_prop_is_special(zfs_prop_t prop)`, similar to `zfs_prop_inheritable()`) before calling `zfs_prop_set_special()` and going through the incidental overhead.

That said, I am looking through the code that uses `zfs_prop_set_special()`, and I see that what you do with these code changes is not out of line with the current uses of `zfs_prop_set_special()`, and the incidental inefficiencies are not addressed elsewhere either. Also, this is probably not performance/resource critical, as property setting is not a frequent/concurrent operation anyway. So, if you don't feel like doing much about this, that's fine with me.

More to the subject matter of you PR, I have experimented a bit with setting inheritable properties across namespace hierarchies that involve zvols, and I found that the way the source of `snapdev` is set is not consistent with other inheritable properties. It appears that simple setting of an inheritable property (e.g. `dedup`) on a top filesystem automagically inherits that value for all the descendants (filesystems and zvols), and the source is marked as such (inherited from top filesystem).  I also found that this behavior changes if one explicitly sets or inherits from some filesystem in the hierarchy - the subsequent changes of top-level property no longer affect the subtree that was inherited explicitly, and the source of inheritance is properly tracked (from specific filesystem and form the top level).

In case of `snapdev` however, setting the top-level filesystem property results in the values propagated but the source  is 'local', not 'inherited from ...'. Also, regardless of explicit setting of `snapdev` in the middle of the hierarchy, the changes to top-level fs are always propagated down the hierarchy, which is not quite how this works for `dedup`.

All of this is easy to see on a toy pool with a few datasets. If it helps, I can attach a terminal session log that illustrates the points made above.

To summarize, as I understand it, there are two issues addressed in this PR:
1) NULL pointer dereference in `zvol_free()`
2) tracking of snapdev property source that is inconsistent with other properties

These two issues are quite separate, and the second does not seem to be fully addressed with your changes. Therefore, I recommend that you split this PR in two, and work separately on addressing the inheritance source/value propagation inconsistencies for `snapdev`.